mlco2/codecarbon
# [Version 3] Add CPU load tracking

When none of RAPL or PowerGadget is available, the mode will use CPU load to estimate the CPU Power.

I've also updated  the CPU database.

I've also added a better default mode : estimate the TDP based on the CPU cores.

How to test:
```
export CODECARBON_LOG_LEVEL=DEBUG
```
To check if it's working : `CPU load 180 W x 90.9% = 163.62 for whole machine.`

If you want to monitor only your process:
```
export CODECARBON_TRACKING_MODE=process
```

Breaking changes:
- The RAPL energy and power reported is now divided by 2 for all "AMD Ryzen Threadripper", following tests on a "AMD Ryzen Threadripper 1950X". Let us know if it doesn't work on other Threadripper. A warning log has been added.

TODO:
- [x] Check that it works when in RAPL mode and file is read protected we switch to this mode.
- [ ] Fix test error "AttributeError: 'OfflineEmissionsTracker' object has no attribute '_country_name'" in [Github Action](https://github.com/mlco2/codecarbon/runs/6869367754?check_suite_focus=true#step:5:663)
- [ ] Add support for process mode (See https://stackoverflow.com/questions/41206809/psutil-measuring-the-cpu-usage-of-a-specific-process )
- Tests to do:
  - [ ] Test on a VM
  - [ ] Test under container LXC
  - [ ] Test under Docker
  - [ ] Test under Windows
  - [x] Test under Linux with RAPL
  - [x] Test under Linux without RAPL
